USB STM32: clean up code and enhancements

kbidani opened the pull request:
This pull request introduces several enhancements and refactorings to the USB code, device tree bindings, and board configurations for STM32 devices. The changes aim to improve the maintainability of the USB driver code.

kbidani committed:
drivers: usb: device_stm32: refactor usb instance assignment

Replace the conditional compilation blocks with a single assignment.
Rely on the DT reg property for the USB controller instance
IOMEM base address.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

kbidani committed:
drivers: usb: udc_stm32: refactor USB instance assignment

Replace the conditional compilation blocks with a single assignment.
Rely on the DT reg property for the USB controller instance
IOMEM base address.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

erwango commented on dts/bindings/usb/st,stm32-otgfs.yaml:
```diff
@@ -31,5 +31,16 @@ properties:
     type: phandle
     description: PHY provider specifier
 
+  phy-type:
+    type: int
+    description: >
+      Type of PHY (Physical Layer) interface used by the OTG_FS/OTG_HS interface.
+
+      The general features of the OTG_FS/OTG_HS interface are as follows:
+
+        - 1: Uses the ULPI interface for an external high-speed PHY for OTG_HS.
+        - 2: Uses the on-chip full-speed PHY for OTG_FS.
+        - 3: Uses the on-chip full-speed PHY for OTG_HS.
```
Let's replace by a string enum. See https://github.com/zephyrproject-rtos/zephyr/pull/82955 for instance.

erwango reviewed (changes requested):
It would be nice to explain in the PR message that you're cleaning up 2 distinct settings:
-Instance > Simple change
-phy interface > Requires introduction of a new prop in the driver

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1018,6 +1018,7 @@ static void priv_pcd_prepare(const struct device *dev)
 	priv->pcd.Init.ep0_mps = cfg->ep0_mps;
 	priv->pcd.Init.speed = PCD_SPEED_FULL;
 	priv->pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
+	priv->pcd.Init.phy_itface = DT_INST_PROP(0, phy_type);
```
This change is not a refactoring, so for the very least, commit message should be updated.
Now, it is very interesting that this was not done on the driver. It is useless (and then maybe it is also useless in the other driver, and then it could just be removed) or was it actually missing here which would have been a bug ?

erwango replied:
Ok, I see that this commit was missing the matching code removal which is not added. Resolved.

tmon-nordic commented:
I would appreciate if the commit messages included stm32 in the title. Initially I thought this PR would be some general UDC interface cleanup, while in reality it tends to cover only stm32.

kbidani commented on dts/bindings/usb/st,stm32-otgfs.yaml:
```diff
@@ -31,5 +31,9 @@ properties:
     type: phandle
     description: PHY provider specifier
 
+  phy-type:
```
done

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -472,24 +472,14 @@ static int usb_dc_stm32_init(void)
-	usb_dc_stm32_state.pcd.Instance = USB_DRD_FS;
-#endif
 	usb_dc_stm32_state.pcd.Init.speed = PCD_SPEED_FULL;
 	usb_dc_stm32_state.pcd.Init.dev_endpoints = USB_NUM_BIDIR_ENDPOINTS;
```
It seems that this line could now be factorized (new commit).

kbidani committed:
drivers: usb: udc_stm32: refactor USB phy_itface assignment

Replace the conditional compilation blocks with a dedicated
function to determine the USB PHY interface selection for STM32
using device tree configurations.
Define generic macros for USB physel that supported by STM32 boards,
while the HALs use different macro names for identical values.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -473,14 +473,14 @@ static int usb_dc_stm32_init(void)
 	int ret;
 	unsigned int i;
 	usb_dc_stm32_state.pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
+	usb_dc_stm32_state.pcd.Init.phy_itface = PHY_ITFACE;
```
This line should go to next commit (or be dropped)

erwango commented on dts/bindings/usb/st,stm32f7-otghs.yaml:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2025, STMicroelectronics
+# SPDX-License-Identifier: Apache-2.0
+
+description: STM32F4 OTGHS controller
```
```suggestion
description: STM32F7 OTGHS controller
```

erwango commented on dts/bindings/usb/st,stm32f4-otghs.yaml:
```diff
@@ -0,0 +1,21 @@
+      - "USB_OTG_HS_ULPI_PHY"
+      - "USB_OTG_FS_EMBEDDED_PHY"
+    description: |
+      ST STM32F4 USB
```
This line could be dropped. Apply everywhere

kbidani committed:
drivers: usb: device_stm32: refactor USB phy_itface assignment

Replace the conditional compilation blocks with a dedicated
function to determine the USB PHY interface selection for STM32
using device tree configurations.
Define generic macros for USB physel that supported by STM32 boards,
while the HALs use different macro names for identical values.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

erwango commented on dts/bindings/usb/st,stm32f7-otghs.yaml:
```diff
@@ -0,0 +1,22 @@
+      - "USB_OTG_HS_EMBEDDED_PHY"
+    description: |
+      OTG_HS supports the following PHY interfaces:
+        - "USB_OTG_HS_ULPI_PHY": an UTMI+ Low Pin Interface for external high-speed PHY.
+        - "USB_OTG_FS_EMBEDDED_PHY": full speed serial transceiver select.
+        - "USB_OTG_HS_EMBEDDED_PHY": an on-chip full-speed PHY.
```
Can we harmonize the wording
```suggestion
        - "USB_OTG_HS_ULPI_PHY": UTMI+ Low Pin Interface for external high-speed PHY.
        - "USB_OTG_FS_EMBEDDED_PHY": on chip full speed PHY.
        - "USB_OTG_HS_EMBEDDED_PHY": on-chip high-speed PHY.
```
Also, on F7, I'm not seeing this `USB_OTG_HS_EMBEDDED_PHY` option (F76x/F77x). Can you cross check?

kbidani replied:
Done.

marwaiehm-st commented on dts/bindings/usb/st,stm32-usb.yaml:
```diff
@@ -31,5 +31,13 @@ properties:
+  physel:
+    type: string
+    enum:
+      - "USB_OTG_FS_PHY_EMBEDDED"
```
should it be set as default like otgfs ?

marwaiehm-st commented on dts/bindings/usb/st,stm32-usb.yaml:
```diff
@@ -31,5 +31,12 @@ properties:
+    type: string
+    default: "USB_OTG_FS_EMBEDDED_PHY"
+    description: |
+      OTG_FS supports the following PHY interfaces:
```
The correct term for the USB controller  is `USB` or `USB_DRD_FS`, not `USB_OTG_FS`. Please update the description accordingly.
```suggestion
      USB supports the following PHY interfaces:
```

kbidani commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -472,33 +472,16 @@ static int usb_dc_stm32_init(void)
 	unsigned int i;
-
+	usb_dc_stm32_state.pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
+	usb_dc_stm32_state.pcd.Init.phy_itface = DT_INST_PROP(0, phy_type);
```
DT_INST_STRING_UPPER_TOKEN used instead: the value of phy_type as an uppercased token, i.e. without any quotes and with special characters converted to underscores.

kbidani commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1017,28 +1017,14 @@ static void priv_pcd_prepare(const struct device *dev)
 	priv->pcd.Init.speed = PCD_SPEED_FULL;
-
+	priv->pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
+	priv->pcd.Init.phy_itface = DT_INST_PROP(0, phy_type);
```
DT_INST_STRING_UPPER_TOKEN used instead: the value of phy_type as an uppercased token, i.e. without any quotes and with special characters converted to underscores.

kbidani commented on dts/bindings/usb/st,stm32-otgfs.yaml:
```diff
@@ -31,5 +31,19 @@ properties:
+    description: |
+      Type of PHY (Physical Layer) interface used by the OTG_FS/OTG_HS interface.
+      The general features of the OTG_FS/OTG_HS interface are as follows:
+        - "USB_OTG_ULPI_PHY": Uses the ULPI interface for an external high-speed PHY for OTG_HS.
```
Done.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -478,13 +478,12 @@ static int usb_dc_stm32_init(void)
 	unsigned int i;
 	usb_dc_stm32_state.pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
 	usb_dc_stm32_state.pcd.Init.phy_itface = DT_INST_STRING_UPPER_TOKEN(0, physel);
+	usb_dc_stm32_state.pcd.Init.dev_endpoints = DT_INST_PROP(0, num_bidir_endpoints);
```
Prefer use `USB_NUM_BIDIR_ENDPOINTS` macro that is defined and used elsewhere in this source file.

kbidani replied:
using DT_INST_PROP(0, num_bidir_endpoints) directly here makes the code more readable. It clearly indicates that the value is sourced from the DT, which is immediately understandable.

kbidani replied:
@erwango what do you think?

kbidani replied:
Rename the macro to DT_USB_NUM_BIDIR_ENDPOINTS to make it clear that the value is sourced from the DT files.

etienne-lms commented on dts/bindings/usb/st,stm32-otgfs.yaml:
```diff
@@ -31,5 +31,13 @@ properties:
     type: phandle
     description: PHY provider specifier
 
+  physel:
```
The property seems `required`. 
Ditto in the other modified DT bindings YAML files.

kbidani replied:
Make the value of physel default.

etienne-lms commented on dts/arm/st/u5/stm32u595.dtsi:
```diff
@@ -101,7 +101,7 @@
 
 		usbotg_hs: otghs@42040000 {
-			compatible = "st,stm32-otghs";
+			compatible = "st,stm32u5-otghs", "st,stm32-otghs";
```
The `physel` propoerty seems required. Should this DT node (and few others in this series where only the compatible property is updated) also set a default `physel` property, possibly overridden by board DTS files?

erwango replied:
Property should indeed be required, but adding default is superfluous if we're populating the property all across the tree (specifically for cases where only one value is possible).
(Also `default` makes `required` less pertinent)

mathieuchopstm reviewed (commented):
General remark: enum values could follow DT `lowercase-dashes` format instead of `UPPERCASE_UNDERSCORES` as done now. (Thanks to `DT_INST_STRING_UPPER_TOKEN`, no modification in the code should be needed)

kbidani commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -40,6 +40,10 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define UDC_STM32_IRQ		DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, irq)
 #define UDC_STM32_IRQ_PRI	DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, priority)
 
+#define USB_OTG_HS_PHY_ULPI	1U
+#define USB_OTG_FS_EMBEDDED_PHY	2U
+#define USB_OTG_HS_EMBEDDED_PHY	3U
```
I will add a comment because the values differ for each board.

kbidani commented on dts/arm/st/f0/stm32f042.dtsi:
```diff
@@ -73,6 +73,7 @@
 			num-bidir-endpoints = <8>;
 			ram-size = <1024>;
 			phys = <&usb_fs_phy>;
+			physel = "USB_OTG_FS_EMBEDDED_PHY";
```
replaced with USB_FS_EMBEDDED_PHY

kbidani commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -65,6 +65,10 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
 
+#define USB_OTG_HS_PHY_ULPI	1U
+#define USB_OTG_FS_EMBEDDED_PHY	2U
+#define USB_OTG_HS_EMBEDDED_PHY	3U
```
Comment added to explain where the magic values ​​come from.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -65,8 +65,9 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
 #define USB_OTG_HS_PHY_ULPI	1U
-#define USB_OTG_FS_EMBEDDED_PHY	2U
+#define USB_FS_EMBEDDED_PHY	2U
 #define USB_OTG_HS_EMBEDDED_PHY	3U
```
These 2 fixup changes should go in commit "drivers: usb: device: refactor usb phy_itface assignment".

kbidani replied:
You're right! 👍

etienne-lms replied:
Changes from line 67 to 70 can be removed, already part of commit "drivers: usb: device: refactor usb phy_itface assignment" where they belong.

kbidani replied:
Done.

etienne-lms replied:
All fine. Thanks.

etienne-lms reviewed (commented):
1 comment to address, otherwise looks all good to me.

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -59,11 +59,16 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
-#define USB_NUM_BIDIR_ENDPOINTS	DT_INST_PROP(0, num_bidir_endpoints)
-#define USB_RAM_SIZE		DT_INST_PROP(0, ram_size)
+#define USB_BASE_ADDRESS		DT_INST_REG_ADDR(0)
+#define USB_IRQ				DT_INST_IRQ_BY_NAME(0, USB_IRQ_NAME, irq)
```
Nit: usually, this is called `IRQN` rather than `IRQ`.

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -472,33 +477,15 @@ static int usb_dc_stm32_init(void)
 	int ret;
 	unsigned int i;
-
+	usb_dc_stm32_state.pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
```
Let's use the macro defined earlier:
```suggestion
	usb_dc_stm32_state.pcd.Instance = (PCD_TypeDef *)USB_BASE_ADDRESS;
```
(Otherwise, it looks like `USB_BASE_ADDRESS` is unused)

etienne-lms replied:
nitpicking: you forgot to remove `USB_BASE_ADDRESS` macro in this source file.

mathieuchopstm commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -757,7 +744,7 @@ int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data * const cfg)
 	}
 
-	if (ep_idx > (USB_NUM_BIDIR_ENDPOINTS - 1)) {
+	if (ep_idx > (DT_USB_NUM_BIDIR_ENDPOINTS - 1)) {
```
Nit: use same form as L194 for consistency
```suggestion
	if (ep_idx >= DT_USB_NUM_BIDIR_ENDPOINTS) {
```

mathieuchopstm reviewed (dismissed):
LGTM with minor non-blockers:

etienne-lms reviewed (commented):
All changes LGTM.

Comment on commit messages:
* commit "dts: bindings: usb: add new compatibles"
and commit "dts: bindings: usb: add new property physel":
Remove sentence *"Also add defines to avoid magic values in the dtsi."* that are added in later commits, straight in the controller drivers source files.

* commit "drivers: usb: device: refactor usb instance assignment"
and commit "drivers: usb: udc: refactor usb instance assignment":
Mention that you now rely on the DT reg property for USB controller instance IOMEM base address.
* commit "drivers: usb: udc: refactor usb phy_itface assignment"
and commit "drivers: usb: device: refactor usb phy_itface assignment":
Mention you define generic macros for USB PHY selection that apply to all STM32 SoC while the HALs uses different macro names for the very same values. 
* commit "drivers: usb: device: refactor endpoint assignement"
s/assignement/assignment/
I think you can rephrase the message body to *"Rename USB_NUM_BIDIR_ENDPOINTS macro and factorize its use in usb_dc_stm32_init()."*

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -59,11 +59,15 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
+#define USB_IRQ				DT_INST_IRQ_BY_NAME(0, USB_IRQ_NAME, irq)
+#define USB_IRQ_PRI			DT_INST_IRQ_BY_NAME(0, USB_IRQ_NAME, priority)
+#define DT_USB_NUM_BIDIR_ENDPOINTS	DT_INST_PROP(0, num_bidir_endpoints)
+#define USB_RAM_SIZE			DT_INST_PROP(0, ram_size)
```
Changes above should go in commit "drivers: usb: udc: refactor usb instance assignment".

kbidani replied:
Here, I deleted the USB_BASE_ADDRESS macro so it's related to this commit.

etienne-lms replied:
The indentation changes from line 62 to line 65 should rather belong to commit
"drivers: usb: device: refactor endpoint assignement" that addresses `USB_NUM_BIDIR_ENDPOINTS` macro changes.

kbidani replied:
Done.

etienne-lms commented:
Nitpicking comment on commit messages for the several commits "drivers: usb: device: refactor ... assignment":
s/assignement/assignment/
(also in header line for the last commit "drivers: usb: device: refactor endpoint assignement"

Some CI build errors seem unrelated:
> /__w/zephyr/zephyr/drivers/i2c/i2c_ll_stm32_v2.c:701: undefined reference to `mem_attr_check_buf'

etienne-lms commented:
The 3 build issues in CI "twister-build (4)" test are related to STM32 I2C driver. P-R https://github.com/zephyrproject-rtos/zephyr/pull/87681 may fix it or at least help it's getting fixed.

The build issue reported by CI "twister-build (17)" seems trickier to me:
> /opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: invalid length for memory region EXTFLASH

It seems related to what is a build warning (in my development env) introduced by bd98e0199b86. I'm looking a it. I'll crated a P-R or an Issue depending on my ability to understand the issue.

etienne-lms commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -472,24 +471,14 @@ static int usb_dc_stm32_init(void)
 	HAL_StatusTypeDef status;
 	int ret;
 	unsigned int i;
-
```
It would be better the empty line is preserved (separates local variable definitions from instructions).

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1017,20 +1017,12 @@ static void priv_pcd_prepare(const struct device *dev)
 	priv->pcd.Init.speed = PCD_SPEED_FULL;
-
+	priv->pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
 	/* Per controller/Phy values */
```
By the way, inline comments at lines 1016 and 1021 no more applies. They can be removed IMHO.
Alternatively `priv->pcd.Instance` can be moved below inlin comment `/* Per controller/Phy values */`.

jfischer-no commented:
@kbidani Please use the git log to see how we write commit messages, especially in USB support. I am moving it to draft until it is fixed.

etienne-lms commented:
FYI #87681 and #87689 have been merged. You can rebase and hopefully have the CI tests running successfully.

kbidani commented:
> jfischer-no

The issue has been fixed. Could you please take a look and confirm if everything is now conform?

erwango commented:
@jfischer-no PTAL

JarmouniA reviewed (approved):
Looks good

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -40,6 +40,11 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define UDC_STM32_IRQ		DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, irq)
 #define UDC_STM32_IRQ_PRI	DT_INST_IRQ_BY_NAME(0, UDC_STM32_IRQ_NAME, priority)
 
+/* Values comes from HAL macros. */
```
What does it mean. The value are actually defined here.

kbidani replied:
The comment was added to clarify that the values being used are derived from the HAL macros. This means that instead of using magic numbers directly in the code, we are now referencing predefined macros from the HAL.

jfischer-no replied:
The comment still does not make any sense. I did not quite understand your explanation. You might want to explain that these definitions are used to match the values defined in the new bindings to get the enum index from the DT. But you should not define anything with a USB_ prefix here, I have complained about this several times in other PRs related to this driver. Also, why not get the index directly from DT?

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -59,10 +59,10 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
-#define USB_RAM_SIZE		DT_INST_PROP(0, ram_size)
+#define USB_IRQ				DT_INST_IRQ_BY_NAME(0, USB_IRQ_NAME, irq)
+#define USB_IRQ_PRI			DT_INST_IRQ_BY_NAME(0, USB_IRQ_NAME, priority)
+#define DT_USB_NUM_BIDIR_ENDPOINTS	DT_INST_PROP(0, num_bidir_endpoints)
```
What is the reason to rename USB_NUM_BIDIR_ENDPOINTS?

kbidani replied:
The macro was renamed to indicate that its value comes from the device tree.

etienne-lms replied:
As far as many other `USB_xxx` macros defined in this source file relate to DT info but are not prefixed with `DT_USB_`, I would tend to agree the renaming is not really needed nor really makes the implementation more consistent.

kbidani replied:
revert.

kbidani commented on dts/bindings/usb/st,stm32f7-otghs.yaml:
```diff
@@ -0,0 +1,22 @@
+    required: true
+    enum:
+      - "USB_OTG_HS_PHY_ULPI"
+      - "USB_FS_EMBEDDED_PHY"
```
The STM32F769I-DISCO board has a USB_OTG_HS_PHY_ULPI. So, I need this node to retrieve the physel and clean up the code to make it easier.

tmon-nordic commented:
Please rebase and fix the conflicts.

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1017,24 +1017,16 @@ static void priv_pcd_prepare(const struct device *dev)
 	priv->pcd.Init.ep0_mps = cfg->ep0_mps;
 	priv->pcd.Init.speed = PCD_SPEED_FULL;
-
+	priv->pcd.Init.phy_itface = DT_INST_STRING_UPPER_TOKEN(0, physel);
```
Remove this line (from commit "drivers: usb: udc_stm32: refactor USB instance assignment").
`physel` porperty is handled in the next commit ("drivers: usb: udc_stm32: refactor USB phy_itface assignment").

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -991,23 +995,12 @@ static void priv_pcd_prepare(const struct device *dev)
+	priv->pcd.Instance = (PCD_TypeDef *)DT_INST_REG_ADDR(0);
+	priv->pcd.Init.phy_itface = DT_INST_STRING_UPPER_TOKEN(0, physel);
 
 #if USB_OTG_HS_EMB_PHY
```
It seems you forgot to remove the lines below in your last P-R update (commit "drivers: usb: udc_stm32: refactor USB phy_itface assignment").

erwango commented:
@jfischer-no PTAL

jfischer-no commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -737,7 +736,7 @@ int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data * const cfg)
 	}
 
-	if (ep_idx > (USB_NUM_BIDIR_ENDPOINTS - 1)) {
+	if (ep_idx >= USB_NUM_BIDIR_ENDPOINTS) {
```
>Rename 'USB_NUM_BIDIR_ENDPOINTS' macro

I do not see any renaming. Why are you making these changes? Summarize it and put it in the commit message.

kbidani replied:
I initially made a change, but then reverted it.

jfischer-no reviewed (changes requested):
Can you please pay more attention to my comments?

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -985,28 +989,43 @@ static const struct udc_stm32_config udc0_cfg  = {
-	priv->pcd.Init.speed = UTIL_CAT(UDC_STM32_, DT_INST_STRING_UPPER_TOKEN(0, maximum_speed));
-
 	/* Per controller/Phy values */
+	priv->pcd.Init.speed = UTIL_CAT(UDC_STM32_, DT_INST_STRING_UPPER_TOKEN(0, maximum_speed));
```
To be done in a dedicated commit ?

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -985,28 +989,43 @@ static const struct udc_stm32_config udc0_cfg  = {
+{
+	uint8_t physel;
+
+#if DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy)
```
Isn't it possible to have this compatible `usb-ulpi-phy` enabled with `usb-nop-xceiv` available as well today ?

kbidani committed:
drivers: usb: device_stm32: refactor endpoint assignment

Factorize 'USB_NUM_BIDIR_ENDPOINTS' macro use in usb_dc_stm32_init().
Modifie the condition to verify ep_idx from > to >=.

Signed-off-by: Khaoula Bidani <khaoula.bidani-ext@st.com>

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -985,28 +989,55 @@ static const struct udc_stm32_config udc0_cfg  = {
+	uint8_t physel;
+
+#if DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy)
+	physel = USB_OTG_HS_PHY_ULPI;
```
I'd add a comment here stating that if "usb-ulpi-phy" is enabled on board.dts, we suppose it is selected (otherwise it doesn't make no sense to define it).

erwango reviewed (commented):
Minor comment, otherwise LGTM

kbidani commented:
@jfischer-no PTAL

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -62,6 +62,10 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #else
 #define UDC_STM32_FULL_SPEED             USB_OTG_SPEED_FULL
 #endif
+/* Definitions aligned with HAL macros for PHY type selection */
+#define USB_OTG_HS_PHY_ULPI	1U
+#define USB_FS_EMBEDDED_PHY	2U
+#define USB_OTG_HS_EMBEDDED_PHY	3U
```
Sorry, maybe I missed the explanations for these. But these macros are not properly prefixed and USB_OTG_HS_EMBEDDED_PHY is already defined in the [HAL](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32f7xx/drivers/include/stm32f7xx_ll_usb.h#L279-L281), would it not clash at some point? Also, I see two other values defined there, why cannot they be used here in the UDC driver?

jfischer-no replied:
>USB_OTG_HS_EMBEDDED_PHY is already defined in the [HAL](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32f7xx/drivers/include/stm32f7xx_ll_usb.h#L279-L281), would it not clash at some point? Also, I see two other values defined there, why cannot they be used here in the UDC driver?

There is one question that remains unanswered. Why cannot the macros from the [HAL](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32f7xx/drivers/include/stm32f7xx_ll_usb.h#L279-L281) be used directly in the UDC driver?

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -985,28 +989,56 @@ static const struct udc_stm32_config udc0_cfg  = {
+
+/* If the ULPI interface for an external high-speed PHY is enabled, it is assumed to be chosen. */
+#if DT_HAS_COMPAT_STATUS_OKAY(usb_ulpi_phy)
+	physel = USB_OTG_HS_PHY_ULPI;
```
`	physel = USB_OTG_ULPI_PHY;` ?

jfischer-no commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -985,28 +989,56 @@ static const struct udc_stm32_config udc0_cfg  = {
+	(DT_PHANDLE(DT_INST(0, st_stm32_usb), phys) == otgfs_phy) || \
+	(DT_PHANDLE(DT_INST(0, st_stm32_usb), phys) == usb_fs_phy)) && \
+	DT_HAS_COMPAT_STATUS_OKAY(usb_nop_xceiv))
+	physel = USB_FS_EMBEDDED_PHY;
```
`	physel = USB_OTG_EMBEDDED_PHY;` ?

mathieuchopstm commented:
Mostly superseded by #96028. 

Feel free to re-open with only the clean-up commits; otherwise, they will be integrated as part of another PR with more cleanups.
